src/main/java/io/lettuce/core/RedisCommandListenersWriter.java

```diff
@@ -0,0 +1,84 @@
+package io.lettuce.core;
+
+import io.lettuce.core.models.events.CommandStartedEvent;
+import io.lettuce.core.protocol.ConnectionFacade;
+import io.lettuce.core.protocol.EventListenerCommand;
+import io.lettuce.core.protocol.RedisCommand;
+import io.lettuce.core.resource.ClientResources;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * @author sokomishalov
+ */
+public class RedisCommandListenersWriter implements RedisChannelWriter {
+
+    private final RedisCommandListener listener;
+    private final RedisChannelWriter delegate;
+
+    public RedisCommandListenersWriter(RedisChannelWriter delegate, RedisCommandListener listener) {
+        this.delegate = delegate;
+        this.listener = listener;
+    }
+
+
+    @Override
+    public <K, V, T> RedisCommand<K, V, T> write(RedisCommand<K, V, T> command) {
+        CommandStartedEvent<K, V, T> event = new CommandStartedEvent<>(command);
+        listener.commandStarted(event);
+
+        return delegate.write(new EventListenerCommand<>(event, listener));
+    }
+
+    @Override
+    public <K, V> Collection<RedisCommand<K, V, ?>> write(Collection<? extends RedisCommand<K, V, ?>> redisCommands) {
+        List<? extends CommandStartedEvent<K, V, ?>> events = redisCommands
+                .stream()
```

Comment:
Using the Stream API imposes severe CPU and memory pressure.

Reply:
fixed